File: web/src/stores/alerts.ts
import { defineStore } from 'pinia'
import { computed, reactive, ref } from 'vue'
import { get, post, put } from '@/api/http'

/** AlertView 是后端的对外表示：**没有任何凭据**，只说「配没配」。 */
export interface AlertView {
  enabled: boolean
  min_level: 'all' | 'warn' | 'crit'
  at_all_on_crit: boolean
  max_retries: number
  webhook_configured: boolean
  lark_configured: boolean
  lark_signed: boolean
  sent: number
  failed: number
  dropped: number
}

type Loader = () => Promise<AlertView>
type Saver = (body: unknown) => Promise<AlertView>
type Tester = () => Promise<unknown>

export const useAlertsStore = defineStore('alerts', () => {
  /**
   * form 里的凭据字段**永远从空开始**。
   *
   * 后端不回显明文（读接口里就没有），所以这里也没有可填回去的东西。
   * 留空提交表示「不改动」——把空当成清空的话，改一次通知级别就会顺手把
   * Webhook 抹掉，而界面上一切正常。
   */
  const form = reactive({
    enabled: false,
    min_level: 'warn' as AlertView['min_level'],
    at_all_on_crit: false,
    max_retries: 2,
    webhook_url: '',
    lark_url: '',
    lark_secret: '',
  })

  const webhookConfigured = ref(false)
  const larkConfigured = ref(false)
  const larkSigned = ref(false)
  const stats = ref({ sent: 0, failed: 0, dropped: 0 })
  const loading = ref(false)
  const saving = ref(false)
  const error = ref('')
  const testing = ref(false)
  const testResult = reactive({ ok: false, msg: '' })

  // 清除标记不进 form：它们是一次性的动作，不是设置项。
  const clears = reactive({ webhook: false, lark: false, lark_secret: false })

  let loader: Loader = () => get<AlertView>('/alerts')
  let saver: Saver = (body) => put<AlertView>('/alerts', body)
  let tester: Tester = () => post<unknown>('/alerts/test')

  /** __setIO / __setTester 只供测试替换网络边界。 */
  function __setIO(l: Loader, s: Saver) {
    loader = l
    saver = s
  }
  function __setTester(t: Tester) {
    tester = t
  }

  function apply(v: AlertView) {
    form.enabled = v.enabled
    form.min_level = v.min_level
    form.at_all_on_crit = v.at_all_on_crit
    form.max_retries = v.max_retries
    webhookConfigured.value = v.webhook_configured
    larkConfigured.value = v.lark_configured
    larkSigned.value = v.lark_signed
    stats.value = { sent: v.sent ?? 0, failed: v.failed ?? 0, dropped: v.dropped ?? 0 }
    // 凭据框保存后清空：留着会让人以为它被回显了，下次再点保存就把
    // 同一个值又发一遍
    form.webhook_url = ''
    form.lark_url = ''
    form.lark_secret = ''
    clears.webhook = clears.lark = clears.lark_secret = false
  }

  async function load() {
    loading.value = true
    error.value = ''
    try {
      apply(await loader())
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  /** payload 省掉未填写的凭据字段，让「留空=不改动」在两侧语义一致。 */
  function payload(): Record<string, unknown> {
    const body: Record<string, unknown> = {
      enabled: form.enabled,
      min_level: form.min_level,
      at_all_on_crit: form.at_all_on_crit,
      max_retries: form.max_retries,
    }
    if (form.webhook_url) body.webhook_url = form.webhook_url
    if (form.lark_url) body.lark_url = form.lark_url
    if (form.lark_secret) body.lark_secret = form.lark_secret
    if (clears.webhook) body.clear_webhook = true
    if (clears.lark) body.clear_lark = true
    if (clears.lark_secret) body.clear_lark_secret = true
    return body
  }

  async function save() {
    saving.value = true
    error.value = ''
    try {
      apply(await saver(payload()))
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      saving.value = false
    }
  }

  function clearWebhook() {
    clears.webhook = true
  }
  function clearLark() {
    clears.lark = true
  }
  function clearLarkSecret() {
    clears.lark_secret = true
  }

  async function sendTest() {
    testing.value = true
    testResult.ok = false
    testResult.msg = ''
    try {
      await tester()
      testResult.ok = true
      testResult.msg = '已发出。没收到就说明地址或凭据不对。'
    } catch (e) {
      // 失败原因原样留在界面上：那是配错了时唯一有用的信息
      testResult.ok = false
      testResult.msg = (e as Error).message
    } finally {
      testing.value = false
    }
  }

  const pendingClears = computed(() => clears.webhook || clears.lark || clears.lark_secret)

  return {
    form, webhookConfigured, larkConfigured, larkSigned, stats,
    loading, saving, error, testing, testResult, pendingClears,
    load, save, sendTest, clearWebhook, clearLark, clearLarkSecret,
    __setIO, __setTester,
  }
})

File: internal/api/alerts.go
package api

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/xltxb/edge_caddy/internal/alert"
)

// alertInput 是告警设置的提交体。
//
// 凭据字段留空表示**不改动**：前端拿不到明文（读接口里就没有），提交时那几个
// 框必然是空的。把空当成清空的话，改一次通知级别就顺手把 Webhook 抹了，
// 而界面上一切正常——直到下次真出事没人收到通知。要清空得用 clear_* 说清楚。
type alertInput struct {
	Enabled     bool   `json:"enabled"`
	MinLevel    string `json:"min_level"`
	WebhookURL  string `json:"webhook_url"`
	LarkURL     string `json:"lark_url"`
	LarkSecret  string `json:"lark_secret"`
	AtAllOnCrit bool   `json:"at_all_on_crit"`
	MaxRetries  int    `json:"max_retries"`

	ClearWebhook    bool `json:"clear_webhook"`
	ClearLark       bool `json:"clear_lark"`
	ClearLarkSecret bool `json:"clear_lark_secret"`
}

// maxRetriesCap 是重试次数的上限。
//
// 卡住是因为重试也是流量：一个 500 的 Webhook 配上 50 次重试，一次节点抖动
// 就是几十个请求，而告警本身早就过时了。
const maxRetriesCap = 5

func (h *handler) alertsReady(c *gin.Context) bool {
	if h.deps.Alerts == nil || len(h.deps.Secret) == 0 {
		fail(c, http.StatusServiceUnavailable, codeInternal, "告警功能未装配")
		return false
	}
	return true
}

// getAlerts 返回**不含任何凭据**的对外表示。
func (h *handler) getAlerts(c *gin.Context) {
	if !h.alertsReady(c) {
		return
	}
	cfg, err := alert.Load(c.Request.Context(), h.deps.Store, h.deps.Secret)
	if err != nil {
		h.failErr(c, err, "读取告警设置失败")
		return
	}
	st := h.deps.Alerts.Stats()
	ok(c, gin.H{
		"enabled": cfg.Public().Enabled, "min_level": cfg.Public().MinLevel,
		"at_all_on_crit":     cfg.Public().AtAllOnCrit,
		"max_retries":        cfg.Public().MaxRetries,
		"webhook_configured": cfg.Public().WebhookConfigured,
		"lark_configured":    cfg.Public().LarkConfigured,
		"lark_signed":        cfg.Public().LarkSigned,
		// 投递计数：失败必须能被发现。静默失败的告警系统比没有告警更糟——
		// 人以为自己被保护着。
		"sent": st.Sent, "failed": st.Failed, "dropped": st.Dropped,
	})
}

// putAlerts 保存设置。响应里同样不回显凭据。
func (h *handler) putAlerts(c *gin.Context) {
	if !h.alertsReady(c) {
		return
	}
	var in alertInput
	if err := c.ShouldBindJSON(&in); err != nil {
		fail(c, http.StatusBadRequest, codeBadInput, "请求体不是合法 JSON")
		return
	}
	level := alert.Level(in.MinLevel)
	switch level {
	case "", alert.LevelAll, alert.LevelWarn, alert.LevelCrit:
	default:
		// 悄悄落一个谁也不认识的值，等于把过滤退化成「全发」或「全不发」，
		// 而界面上显示的是用户填的那个词
		fail(c, http.StatusUnprocessableEntity, codeBadInput,
			"通知级别只能是 all / warn / crit，收到 "+in.MinLevel)
		return
	}
	if in.MaxRetries < 0 || in.MaxRetries > maxRetriesCap {
		fail(c, http.StatusUnprocessableEntity, codeBadInput,
			"重试次数须在 0 到 5 之间——重试也是流量，而告警拖久了就没用了")
		return
	}

	next := alert.Config{
		Enabled: in.Enabled, MinLevel: level,
		WebhookURL: in.WebhookURL, LarkURL: in.LarkURL, LarkSecret: in.LarkSecret,
		AtAllOnCrit: in.AtAllOnCrit, MaxRetries: in.MaxRetries,
		ClearWebhook: in.ClearWebhook, ClearLark: in.ClearLark, ClearLarkSecret: in.ClearLarkSecret,
	}
	if err := alert.Merge(c.Request.Context(), h.deps.Store, h.deps.Secret, next); err != nil {
		h.failErr(c, err, "保存告警设置失败")
		return
	}
	saved, err := alert.Load(c.Request.Context(), h.deps.Store, h.deps.Secret)
	if err != nil {
		h.failErr(c, err, "读取告警设置失败")
		return
	}
	// 落库成功才让运行中的通知器换配置：反过来的话，一次写库失败会留下
	// 「内存里已生效、重启后回退」的不一致。
	h.deps.Alerts.SetConfig(saved)
	h.log.Info("告警设置已更新", "operator", operatorOf(c),
		"enabled", saved.Enabled, "min_level", saved.MinLevel)
	// 回的是对外表示，连保存响应里也没有凭据——那是最容易被忽略的一处回显
	ok(c, saved.Public())
}

// testAlerts 发一张测试卡片，同步等结果。
func (h *handler) testAlerts(c *gin.Context) {
	if !h.alertsReady(c) {
		return
	}
	cfg, err := alert.Load(c.Request.Context(), h.deps.Store, h.deps.Secret)
	if err != nil {
		h.failErr(c, err, "读取告警设置失败")
		return
	}
	h.deps.Alerts.SetConfig(cfg)

	if err := h.deps.Alerts.SendTest(c.Request.Context()); err != nil {
		// 成败都写审计（中间件按状态码判定）：只记成功的话，
		// 「我明明测过」与「测了但没通」在事后无法区分。
		h.log.Warn("告警测试发送失败", "operator", operatorOf(c), "err", err)
		fail(c, http.StatusBadGateway, codeInternal, "测试发送失败："+err.Error())
		return
	}
	h.log.Info("告警测试已发送", "operator", operatorOf(c))
	ok(c, gin.H{"sent": true})
}
